File: jump_game.java
public class Solution {
    public boolean canJump(int[] nums) {
        int reach = 0;
        int n = nums.length;
        for (int i = 0; i < n && i <= reach; i++) {
            reach = Math.max(i + nums[i], reach);
        }
        return reach >= nums[n-1]; // or i == n;
    }
}

File: jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int i = 0, n = nums.size(), reach = 0;
        for (int reach = 0; i < n && i <= reach; i++) {
            reach = max(reach, i + nums[i]);
        }
        return i == n;
    }
};
